### PR TITLE
Add support for variables

### PR DESCRIPTION
**Pull request by BenediktSeidl**
Requested in #28 and [feature-295540](https://features.checkmk.com/suggestions/295540/).

You should now be able to create query variables, and use them in the query editor. This feature is fully functional in all checkmk editions.

### CODE REVIEW

File: src/backend/rest.ts
Comment (TheSpacyCat) [inline src/backend/rest.ts:90]: Can we factor these together? I'm thinking of a `RestApiCollectionResponse<ExtenionType>` interface.

File: src/backend/rest.ts
Comment (TheSpacyCat) [inline src/backend/rest.ts:143]: These three could maybe be implemented by calling a common utility method, I think. Their structure is basically the same.

I'm also not sure about the names, at least the first one. It doesn't do any filtering, doesn't it?

File: src/backend/web.ts
Comment (TheSpacyCat) [inline src/backend/web.ts:25]: Note to self: Look at this again later

File: src/ui/VariableQueryEditor.tsx
Comment (TheSpacyCat) [inline src/ui/VariableQueryEditor.tsx:68]: Do we have to have a description?

File: src/ui/components.tsx
Comment (TheSpacyCat) [inline src/ui/components.tsx:48]: I'm not sure about an optional boolean prop, at least `true | undefined` is better, if typescript can do boolean constant type like it can with strings.

Otherwise, we should document a default value.

File: src/ui/components.tsx
Comment (TheSpacyCat) [inline src/ui/components.tsx:84]: That means, this branch should run if `showVariables` is undefined, i.e. left out when the component is instantiated? I'm sure about the logic here.

File: src/ui/components.tsx
Comment (TheSpacyCat) [inline src/ui/components.tsx:133]: The second argument seems unnecessary to me. Shouldn't typescript be happy about it being left out?

File: src/utils.ts
Comment (TheSpacyCat) [inline src/utils.ts:143]: I'd like a for-loop more here.

File: src/utils.ts
Comment (TheSpacyCat) [inline src/utils.ts:175]: What error could happen here? Should we inform the user?

File: src/DataSource.ts
Comment (TheSpacyCat) [inline src/DataSource.ts:44]: In the future, we should think about moving the `filter*` functions to a class outside the backends. The Datasource shouldn't be the one to decide where to get its data from.

**Review by TheSpacyCat — CHANGES_REQUESTED** [review]

**Review by TheSpacyCat — COMMENTED** [review]
Looks a bit rough, but polishing can wait until after the release, I guess.

File: src/ui/QueryEditor.tsx
Comment (BenediktSeidl) [reply src/ui/QueryEditor.tsx:135]: VerticalGroup will wrap them as one element when both share the same variable. but it works if you use two variables for each select.

File: src/ui/components.tsx
Comment (BenediktSeidl) [reply src/ui/components.tsx:433]: we have this on our todo list, so i'm going to resolve it here.

File: src/backend/rest.ts
Comment (BenediktSeidl) [reply src/backend/rest.ts:143]: yes, i also thought about combining them, but my final thought was that the differences are big enought that the resulting common function is more complicate than the duplicated independent one.
i would rename them, but keep them as is. if you have objection feel free to reopen the discussion!

File: src/ui/components.tsx
Comment (BenediktSeidl) [reply src/ui/components.tsx:133]: seems to work without

File: src/utils.ts
Comment (BenediktSeidl) [reply src/utils.ts:143]: why? please sell it to me :-) i don't see a problem here.
i think i implemented it for better typing support, and i fear that this typing support is going to vanish when changed to a for loop.

File: src/utils.ts
Comment (BenediktSeidl) [reply src/utils.ts:175]: this looks like a debugging artifact, i removed it.

File: tests/unit/ui/CheckMkAsyncSelect.test.tsx
Comment (BenediktSeidl) [reply tests/unit/ui/CheckMkAsyncSelect.test.tsx:15]: can you do this in an additional commit, after we merged this into main? i'm not familiar with jest and it would take quite some time for me.

File: src/DataSource.ts
Comment (BenediktSeidl) [reply src/DataSource.ts:44]: added to the todo list

File: src/ui/components.tsx
Comment (BenediktSeidl) [reply src/ui/components.tsx:48]: setting it to `true | undefined` make this property non optional, so you would have to explicitly set it to undefined. then i guess a normal boolean would be better.
i changed it to `showVariables?: false` and added a comment.

File: src/ui/components.tsx
Comment (BenediktSeidl) [reply src/ui/components.tsx:84]: added a comment at the prop definition for clarification.
most of the cases the variables should be shown, but there are some exceptions where `showVariables` is explicitly set to false.

File: src/ui/components.tsx
Comment (BenediktSeidl) [reply src/ui/components.tsx:118]: i think we already have enough hacks, so i would pass here.
i saw also some solution where they listen to browser history changes that are issued by grafana when changing options. still a hack in my opinion. i think listing the variables is already a good benefit, seeing the content would be cool, but is not a must-have.

File: src/utils.ts
Comment (TheSpacyCat) [reply src/utils.ts:143]: This is probably only about what I'm used to, but when I see a `forEach`, is something that is done for side effects alone, while here we're producing another value in `result`. So I'll change my opinion now and say that a `map` over `Object.entries(requestSpec)` would be more appropriate here, but I guess that would make typing even harder.

I'm not sure what type information you would lose with a for loop here. Your loop variable would have type `keyof Partial<RequestSpec>` and the rest should be the same.

Anyway, I don't see this as a blocker. If rewriting would be too much effort, I'd be happy to think about it the next time I stumble across it.

File: src/DataSource.ts
Comment (TheSpacyCat) [inline src/DataSource.ts:37]: `unknown` instead of `any` would be fine here, since we're not using the parameter.

**Review by TheSpacyCat — APPROVED** [review]

**Pull request merged**
